### ISSUE TITLE
MergeWindows - Place popups next to their openers not working

### ISSUE THREAD

**Issue by onemen** [onemen/TabMixPlus#64]
https://github.com/onemen/TabMixPlus/blob/316280c16adf849821aa6c8283eafdd76dd66f2b/addon/modules/MergeWindows.jsm#L135

@117649 ,

Old code to **Place popups next to their openers** does not work.
1. the function gBrowser._getTabForContentWindow was removed.
2. "Tabmix:collectOpener" content message listener does not work.

using browser.ownerGlobal.opener return ChromeWindow and i don't know how to  get the opener tab from it
let me know if you can find solution to this issue.

I can also remove this option and just merge popup window at the end

**Comment by 117649** [comment]
@onemen Sorry I never used this function. Care to explain what it should do? I'll see what I can find.

**Comment by onemen** [comment]
The function swapTabs(aWindow, tabs) in MergeWindows.jsm uses swapBrowsersAndCloseOther to move tabs to aWindow.
when the tab is popup it look for it opener tab and if it find one it move the popup tab to be after the opener tab in the window

it the past as i mention above i used content message listener to retrieve the opener and sent it back.

**Comment by 117649** [comment]
There seems to be a clear route for replace **gBrowser._getTabForContentWindow** 
: https://hg.mozilla.org/mozilla-central/rev/1940595cb776790de40008a4b90d54356f5e5753

**Comment by onemen** [comment]
the tricky part is to find the contentWindow for the opener 
this does not work
```
let openerWindow = browser.ownerGlobal.opener;
let openerTab = tabbrowser.getTabForBrowser(openerWindow.docShell.chromeEventHandler);
```

**Comment by 117649** [comment]
Any idea why is this?
>     2. "Tabmix:collectOpener" content message listener does not work.


When I search firefox source I found that Mozilla give **sendAsyncMessage()** at least an empty object.
I can't test it yet but may be worth a try?
: https://searchfox.org/mozilla-central/source/browser/components/sessionstore/SessionStore.jsm#5655

**Comment by 117649** [comment]
> 
> 
> Any idea why is this?
> 
> > ```
> > 2. "Tabmix:collectOpener" content message listener does not work.
> > ```

@onemen I'm getting this:
https://github.com/onemen/TabMixPlus/blob/316280c16adf849821aa6c8283eafdd76dd66f2b/addon/chrome/content/scripts/content.js#L140-L141
**sendSyncMessage()** not sending **{opener: content.opener}** it seems only accept two argument.
But for some reason **content.opener** can't be send neither since it is a window and window have a **self** property.
**sendSyncMessage()** seems doing JSON.stringify() and thus fail with "TypeError: cyclic object value".

**Comment by onemen** [comment]
> 
> 
> Any idea why is this?
> 
> > ```
> > 2. "Tabmix:collectOpener" content message listener does not work.
> > ```

In all my testing i get that content.opener is always null

**Comment by 117649** [comment]
> 
> 
> > Any idea why is this?
> > > ```
> > > 2. "Tabmix:collectOpener" content message listener does not work.
> > > ```
> 
> In all my testing i get that content.opener is always null

![图片](https://user-images.githubusercontent.com/8281345/129447540-442ab652-b0e4-4d3a-b62d-988301d46bd0.png)

This one is from : https://javascript.info/popup-windows#example-a-minimalistic-window

**Comment by onemen** [comment]
how you run the example? when i'v tried it from Browser console it throw an error

**Comment by 117649** [comment]
> 
> 
> how you run the example? when i'v tried it from Browser console it throw an error

![图片](https://user-images.githubusercontent.com/8281345/129447907-38e7aa3e-0292-400e-8b28-a89ed62ef505.png)

**Comment by onemen** [comment]
> 
> 
> > how you run the example? when i'v tried it from Browser console it throw an error
> 
> ![图片](https://user-images.githubusercontent.com/8281345/129447907-38e7aa3e-0292-400e-8b28-a89ed62ef505.png)

I don't understand ?

**Comment by 117649** [comment]
> I don't understand ?

I just hit that button and let it open a popup for me...

**Comment by onemen** [comment]
Thank :)

now i get an error is Utils.jsm
```
Utils.jsm:64:9
16:43:19.353 TypeError: can't access property "opener", message.objects is undefined
    receiveMessage chrome://tabmix-resource/content/Utils.jsm:64
```

**Comment by 117649** [comment]
> 
> 
> Thank :)
> 
> now i get an error in Utils.jsm
> 
> ```
> Utils.jsm:64:9
> 16:43:19.353 TypeError: can't access property "opener", message.objects is undefined
>     receiveMessage chrome://tabmix-resource/content/Utils.jsm:64
> ```

Well, that is because:

> https://github.com/onemen/TabMixPlus/blob/316280c16adf849821aa6c8283eafdd76dd66f2b/addon/chrome/content/scripts/content.js#L140-L141
> 
> **sendSyncMessage()** not sending **{opener: content.opener}** it seems only accept two argument.
> But for some reason **content.opener** can't be send either since it is a window and window have a **self** property.
> **sendSyncMessage()** seems doing JSON.stringify() and thus fail with "TypeError: cyclic object value".

**Comment by onemen** [comment]
Maybe we can send the opener window id (if i can find it) and then use gBrowser.getBrowserForOuterWindowID(id) in Utils.jsm

**Comment by 117649** [comment]
> 
> 
> Maybe we can send the opener window id (if i can find it) and then use gBrowser.getBrowserForOuterWindowID(id) in Utils.jsm

**content.opener.docShell.outerWindowID** exist!
![图片](https://user-images.githubusercontent.com/8281345/129449544-833e8833-efb7-48ff-acb4-25a40ba3f66a.png)
![图片](https://user-images.githubusercontent.com/8281345/129449531-ab1f7347-b574-49a4-8dd3-a96a4b200d3a.png)
Works.

**Comment by onemen** [comment]
Yes,

know i need to figure how to get outerWindowID for non-remote popup (is it possible??) and test if changing in this in swapTabs will work
```diff
- openerWindow = browser.contentWindow.opener
+ openerWindowID = browser.contentWindow.opener.docShell.outerWindowID
```

**Comment by onemen** [comment]
It look like we don't need to use content script and get the outerWindowID directly from the browser
```
tab.linkedBrowser.ownerGlobal.top.docShell.outerWindowID
```

**Comment by onemen** [comment]
it looks like we don't need to use content script at all
```
openerWindowID = browser.browsingContext?.opener?.currentWindowGlobal?.outerWindowId;
```
this expression always return the opener outerWindowId

**Comment by onemen** [comment]
Fixed by 32b3de03f1cea46a1403b31a4b4dd9f25741a069
